### ISSUE TITLE
Resampling an image to a reference image

### ISSUE THREAD

**Issue by ctorti** [SimpleITK/SimpleITK#1245]
Hi all,

I'm trying to resample an image (`Image`) using the grid of another image (`RefImage`).  Not only are the slices in the resampled image (`ResImage`) not anatomically related to the slices in `RefImage`, but I can't be sure that the intrically determined `outputSpacing` is correct, since I obtain an incorrect spacing along the third dimension for `RefImage`, i.e. `RefImage.GetSpacing()[2]`. 

I've gone through the [Transforms and Resampling tutorial](https://hub.gke2.mybinder.org/user/insightsoftware-leitk-notebooks-ck6imqqu/notebooks/Python/21_Transforms_and_Resampling.ipynb) and searched for relevant issues already posted by other users but I've been unable to figure out where I'm going wrong.  

I'm looking to use the second method covered in the tutorial under **Resampling**:

> 1. ```Resample(const Image &image1, Transform transform, InterpolatorEnum interpolator, double defaultPixelValue, PixelIDValueEnum outputPixelType)```
> 2. ```Resample(const Image &image1, const Image &referenceImage, Transform transform, InterpolatorEnum interpolator, double defaultPixelValue, PixelIDValueEnum outputPixelType)```
> 3. ```Resample(const Image &image1, std::vector< uint32_t > size, Transform transform, InterpolatorEnum interpolator, std::vector< double > outputOrigin, std::vector< double > outputSpacing, std::vector< double > outputDirection, double defaultPixelValue, PixelIDValueEnum outputPixelType)```

I also found [this helpful example](https://gist.github.com/zivy/79d7ee0490faee1156c1277a78e4a4c4) which looks relevant.  However it seems that zivy was interested in aligning the two images to their geometric centres, whereas I am interested in aligning to their origins*.   So I skipped the contents of the third cell (setting up `reference_center`).  That said, I also tried using the centered transform as shown in that example, but still didn't yield the results I expected. 

*Note: 
The two volumetric images that I have are of the same patient imaged with the same *ImageOrientationPatient* (bar rounding differences), but different *SliceThickness*, *Rows* and *Columns*, and hence, different *ImagePositionPatient*s (using DICOM parlance).   Their origins (= *ImagePositionPatient*[0]) are also different (using SimpleITK parlance).  Hence why I want to align their origins rather than geometric centres.

The code using `Resample`:

```
Dimension = Image.GetDimension()

Transform = sitk.AffineTransform(Dimension)
Transform.SetMatrix(Image.GetDirection())
Transform.SetTranslation(np.array(Image.GetOrigin()) - RefImage.GetOrigin())

ResImage = sitk.Resample(Image, RefImage, Transform, sitk.sitkLinear, Image.GetPixelIDValue())
```

Following are some results printed to the console:

> Results after resampling Source image:
> 
> The Source and Target image size:
>  
>    Original Source:  (208, 256, 50) 
>  
>    Resampled Source: (192, 192, 63) 
>  
>    Target:           (192, 192, 63)
> 
> The Source and Target voxel spacings:
>  
>    Original Source:  (0.8984375, 0.8984375, 2.9999990463256836) 
>  
>    Resampled Source: (1.3020833730698, 1.3020833730698, 1.0) 
>  
>    Target:           (1.3020833730698, 1.3020833730698, 1.0)
> 
> The Source and Target Origin:
>  
>    Original Source:  (-103.26176452636719, -153.43377685546875, -10.299073219299316) 
>  
>    Resampled Source: (-134.8488006591797, -162.71725463867188, -6.488883972167969) 
>  
>    Target:           (-134.8488006591797, -162.71725463867188, -6.488883972167969)
> 
> The Source and Target Direction:
>  
>    Original Source:  (0.9972648738498597, 1.8634260454120088e-10, 0.07391056342109259, -0.022262997957353512, 0.9535561337886633, 0.3003915089278791, -0.07047787104598309, -0.301215370979001, 0.9509480374756598) 
>  
>    Resampled Source: (0.9972648742807271, -8.267604392764305e-09, 0.07391055760745978, -0.022262987236672742, 0.9535561377645052, 0.3003914971015985, -0.07047786833570838, -0.3012153583926953, 0.9509480416632733) 
>  
>    Target:           (0.9972648742807271, -8.267604392764305e-09, 0.07391055760745978, -0.022262987236672742, 0.9535561377645052, 0.3003914971015985, -0.07047786833570838, -0.3012153583926953, 0.9509480416632733)

Unlike the use of `ResampleImageFilter`, this method of `Resample` doesn't require setting `outputSpacing` (as required for `Resample` method 3), presumably since the necessary info is intrinsically obtained from the arguments `image1` and `referenceImage`.  But I get the wrong pixel spacing along the third dimension for `RefImage` (it seems fine for `Image`) as reported by `.GetSpacing()`.  The *SliceThickness* for `Image` and `RefImage` as reported by Pydicom are 3.0 mm and 5.0 mm, yet SimpleITK reports 2.999... mm and 1.0 mm for `.GetSpacing()[2]`.  So if `Resample` is using the wrong `outputSpacing` could this be (at least one of) the issues?

I wondered if this is contributing to the fact that the resulting resampled image is not showing comparable anatomy when I print the slices in `ResImage` and `RefImage` side-by-side.  It's only the middle slice 31 where the anatomy shown in `ResImage` seems to agree with that in `Image`.

I decided to try using `ResampleImageFilter`, both using the voxel spacings given by `RefImage.GetSpacing()` and using the spacings from `Dicom.PixelSpacing` and `Dicom.SliceThickness` with Pydicom (where `Dicom` is a DICOM object imported with Pydicom).  

The code using `ResampleImageFilter`:

```
Dimension = Image.GetDimension()
    
Resampler = sitk.ResampleImageFilter()
Resampler.SetReferenceImage(RefImage)
Resampler.SetInterpolator(sitk.sitkLinear)
Resampler.SetTransform(sitk.AffineTransform(Dimension))

Resampler.SetOutputSpacing(RefImageSpacings) # <-- allow use of Pydicom spacing
Resampler.SetSize(RefImage.GetSize())
Resampler.SetOutputDirection(RefImage.GetDirection())
Resampler.SetOutputOrigin(RefImage.GetOrigin())
Resampler.SetDefaultPixelValue(Image.GetPixelIDValue())

ResImage = Resampler.Execute(Image)
```

where `RefImageSpacings` is either `RefImage.GetSpacing()` or `[Dicom.PixelSpacing[0], Dicom.PixelSpacing[1], Dicom.SliceThickness]`.

Although I get the expected dimensions in `RefImage`, when plotting the slices side-by-side the results appear worse than that obtained using `Resample`.  When I set  `RefImageSpacings = RefImage.GetSpacing()` I get 63 non-zero slices in `ResImage`, but even the central slices of `ResImage` and `RefImage` do not correspond anatomically.  

When I set `RefImageSpacings = [Dicom.PixelSpacing[0], Dicom.PixelSpacing[1], Dicom.SliceThickness]` I get completely black slices from slice number 30 to 62 in `ResImage`.  So it appears that my attempts to correct the error in `outputSpacing` has not only not helped but made the situation worse.

This all leads me to suspect that I'm not using `ResampleImageFilter` properly (either), and hence, I'm no closer to determining whether or not the error in `RefImage.GetSpacing()[2]` is a cause for concern or if other mistakes are to blame for the lack in agreement between `ResImage` and `RefImage`.

I would greatly appreciate if someone could shed some light on this.

**Comment by zivy** [comment]
Hello @ctorti,

The problem isn't with the code, it is with the interpretation of the DICOM tag. [Slice Thickness](https://dicom.innolitics.com/ciods/ct-image/image-plane/00180050), "0018|0050", doesn't mean what you think it means. It is the thickness of the region in space represented by this slice where the infinity thin slice is at the center of this region.

The z spacing is computed based on the [Image Position (Patient)](https://dicom.innolitics.com/ciods/ct-image/image-plane/00200032), "0020|0032",  and [Image Orientation (Patient)](https://dicom.innolitics.com/ciods/ct-image/image-plane/00200037), "0020|0037", tags for the first two slices and assumed to be uniform.

**Comment by ctorti** [comment]
Hi @zivy ,

Thanks for your reply.  I follow that Slice Thickness is the region of space represented by the slice of finite thickness, so is it also not true that the Slice Thickness is equivalent to the voxel dimension along the scan direction?

I did some playing around with Pydicom and SimpleITK to see how they differ.  First I create a list `DicomFpaths` which contain a list of filepaths of DICOMs in a directory that have been sorted by ImagePositionPatient[2].  I import the DICOM objects using Pydicom so I have a list of objects `Dicoms`.  I can then print the list of IPPs in ascending order of IPP[2].

Using SimpleITK I create a list of IPPs using `.TransformIndexToPhysicalPoint((0, 0, i))` for each slice in the sitk image.

As expected when I compare the list of IPPs obtained with Pydicom and SimpleITK they are in agreement for one DICOM scan (the scan that is imported in sitk as `Image`).  To my surprise when I did the same for the other scan (the one that was imported in sitk as `RefImage`) I got repeated positions in the list obtained with Pydicom:

```
[[-134.84880591119, -162.71725421062, -6.4888838763119],
 [-134.84880591119, -162.71725421062, -6.4888838763119],
 [-134.84880591119, -162.71725421062, -6.4888838763119],
 [-134.29447603906, -160.46432070843, 0.64322671939124],
 [-134.29447603906, -160.46432070843, 0.64322671939124],
 [-134.29447603906, -160.46432070843, 0.64322671939124],
...
```

and the differences in the z-components of the IPPs are:

`[0.        , 0.        , 7.1321106 , 0.        , 0.        ,        7.1321106 , ...`

yet unique positions reported by sitk:

```
[[-134.8488006591797, -162.71725463867188, -6.488883972167969],
 [-134.77489010157223, -162.41686314157027, -5.537935930504696],
 [-134.70097954396476, -162.11647164446867, -4.586987888841422],
 [-134.6270689863573, -161.81608014736707, -3.6360398471781488],
 [-134.55315842874984, -161.51568865026547, -2.6850918055148756],
 [-134.47924787114238, -161.21529715316387, -1.734143763851602],
...
```

and the differences in the z-components of the IPPs are:

`[0.95094804, 0.95094804, 0.95094804, 0.95094804, 0.95094804, 0.95094804, ...`

I checked `DicomFpaths` for the second scan and verified that none of the filepaths were duplicated.  The SOPInstanceUIDs were also unique.  

Then I loaded the DICOMs in DicomBrowser and noticed that there are only 21 IPPs listed (there are 63 DICOM files).  Then I noticed that the scan title is "6-ep2ddiff3scantracep2-55446", so it appears that this collection of images have been subtracted from one another in threes.  This explains why I get repeated IPPs from pydicom.  

I note with interest your comment:

> The z spacing is computed based on the Image Position (Patient), "0020|0032", and Image Orientation (Patient), "0020|0037", tags for the first two slices and assumed to be uniform.

I'm surprised because the list of files that I provided to `ImageSeriesReader` was the same ordering used to import the DICOMs in pydicom:

```
SortedFilePaths = GetDicomFpaths(DicomDir, 'slices', False)

SitkReader = sitk.ImageSeriesReader()
SitkReader.SetFileNames(SortedFilePaths)
    
Image = SitkReader.Execute()
Image = sitk.Cast(Image, sitk.sitkFloat32)
```

where `GetDicomFpaths` is a function that creates a list of filepaths in ascending order of IPP[2].  So if the first two IPPs in the list `SortedFilePaths` are the same, shouldn't sitk have reported `Image.GetSize()[2] = 0`?

If you're wondering why I've gone through the trouble of sorting the filepaths is because when I imported the image as follows:

```
SitkReader = sitk.ImageSeriesReader()
SitkNames = SitkReader.GetGDCMSeriesFileNames(DicomDir)
    
SitkReader.SetFileNames(SitkNames)
Image = SitkReader.Execute()
Image = sitk.Cast(Image, sitk.sitkFloat32)
```

and plotted the slices in `Image` it was clear that the slices were not in anatomical order.

So it looks like sitk is capable of loading images in a sensible order if the scan has "sensible" IPPs.  Because I tried to load a scan containing odd IPPs, sitk quite reasonably didn't know how to deal with it when working out `Image.GetSize()` it seems.

In any case it's fairly clear to me that I need to work with a different data set for `RefImage` because I don't know what has been done to the DICOMs in the scan "6-ep2ddiff3scantracep2-55446". 

I'll see if I can obtain better resampling results with two "normal" scans.
